conda/core/index.py

```diff
@@ -52,104 +57,186 @@ def get_index(channel_urls=(), prepend=True, platform=None,
+    from .repodata import collect_all_repodata_as_index
+    index = collect_all_repodata_as_index(use_cache, channel_urls)
 
     return index
 
 
-def _supplement_index_with_prefix(index, prefix, channels):
-    # type: (Dict[Dist, IndexRecord], str, Set[canonical_channel]) -> None
+def _supplement_index_with_prefix(index, prefix):
     # supplement index with information from prefix/conda-meta
     assert prefix
-    maxp = len(channels) + 1
-    for dist, info in iteritems(linked_data(prefix)):
+    for dist, prefix_record in iteritems(linked_data(prefix)):
         if dist in index:
             # The downloaded repodata takes priority, so we do not overwrite.
             # We do, however, copy the link information so that the solver (i.e. resolve)
             # knows this package is installed.
-            old_record = index[dist]
-            link = info.get('link') or EMPTY_LINK
-            index[dist] = IndexRecord.from_objects(old_record, link=link)
+            current_record = index[dist]
+            link = prefix_record.get('link') or EMPTY_LINK
+            index[dist] = PrefixRecord.from_objects(current_record, prefix_record, link=link)
         else:
-            # If the package is not in the repodata, use the local data. If
-            # the 'depends' field is not present, we need to set it; older
-            # installations are likely to have this.
-            depends = info.get('depends') or ()
+            # If the package is not in the repodata, use the local data.
             # If the channel is known but the package is not in the index, it
             # is because 1) the channel is unavailable offline, or 2) it no
             # longer contains this package. Either way, we should prefer any
             # other version of the package to this one. On the other hand, if
             # it is in a channel we don't know about, assign it a value just
             # above the priority of all known channels.
-            priority = MAX_CHANNEL_PRIORITY if dist.channel in channels else maxp
-            index[dist] = IndexRecord.from_objects(info, depends=depends, priority=priority)
+            index[dist] = prefix_record
 
 
-def _supplement_index_with_cache(index, channels):
-    # type: (Dict[Dist, IndexRecord], Set[canonical_channel]) -> None
+def _supplement_index_with_cache(index):
     # supplement index with packages from the cache
-    maxp = len(channels) + 1
-    for pc_entry in PackageCache.get_all_extracted_entries():
-        dist = Dist(pc_entry)
+    for pcrec in PackageCache.get_all_extracted_entries():
+        dist = Dist(pcrec)
         if dist in index:
             # The downloaded repodata takes priority
-            continue
-        pkg_dir = pc_entry.extracted_package_dir
-        index_json_record = read_index_json(pkg_dir)
-        # See the discussion above about priority assignments.
-        priority = MAX_CHANNEL_PRIORITY if dist.channel in channels else maxp
```

Comment:
Same question here, but I think the answer is easier. In this case, one or more of the packages in the cache are not found in any of the requested channels.